File: Fisher-Yates/Go/jcla1/fisher_yates.go
package fisherYates

import (
	"math/rand"
)

// Make sure you seed the PRNG before using this function
// Fisher-Yates shuffle runs in O(n) and uses O(1) space
func FisherYates(arr []int) []int {
	m := len(arr)
	var i int

	for m != 0 {
		i = rand.Intn(m)
		m -= 1
		arr[i], arr[m] = arr[m], arr[i]
	}

	return arr
}

File: Fisher-Yates/Ruby/jcla1/fisher_yates.rb
def fisher_yates(arr)
    m = arr.length
    while m != 0
        i = rand(m)
        m -= 1
        arr[i], arr[m] = arr[m], arr[i]
    end

    return arr
end

# arr = Proc.new { (1..3).to_a }
# d = Hash.new(0)
# 1000000.times { d[fisher_yates(arr.call).join ""] += 1 }
# p d